### ISSUE TITLE
[`module rec`] does not initialize let bindings

### ISSUE THREAD

**Issue by xperiandri** [dotnet/fsharp#16009]
**Repro steps**

Run reproduction sample
[ModuleRecIssueRepro.zip](https://github.com/dotnet/fsharp/files/12673625/ModuleRecIssueRepro.zip)

Notice that in the file `AssessmentDataValidated.fs` at line 61 `Codec` value is `null`!

``` F#
[<AutoOpen>]
module CodecsExtensions =

    module AsseseeCodecs =

        open System.Net.Mail

        let Codec: Codec<Encoding, Assessee> = // <<----------------- null is here!
            ofObjCodec
            <| codec {
                let! lastName = jreq "lastName" (fun a -> a.LastName |> Some)
                and! firstName = jreq "firstName" (fun a -> a.FirstName |> Some)

                and! middleName =
                    jopt "middleName" (fun a -> a.MiddleName |> ValueOption.toOption)

                and! dateOfBirth = jopt "dateOfBirth" (fun a -> a.DateOfBirth |> ValueOption.toOption)
                and! phone = jopt "phone" (fun a -> a.Phone |> ValueOption.toOption)
                and! email = jopt "email" (fun a -> a.Email |> ValueOption.map (fun mail -> mail.Address) |> ValueOption.toOption)

                return {
                    LastName = lastName
                    MiddleName = middleName |> Option.toVOption
                    FirstName = firstName
                    DateOfBirth = dateOfBirth |> Option.toVOption
                    Phone = phone |> Option.toVOption
                    Email = email |> Option.toVOption |> ValueOption.map MailAddress
                }
            }

    type Assessee with

        static member Codec = AsseseeCodecs.Codec

[<RequireQualifiedAccess>]
module Codecs =

    let V1: Codec<Encoding, EventData> =
        let assesseeCodec = Assessee.Codec
        ofObjCodec
        <| codec {
            let! assessee = jreqWith Assessee.Codec "assessee" (fun di -> di.Assessee |> Some)
            let x = { Assessee = assessee(*; DrivingLicense = drivingLicense; Address = address*) }
            return x
        }

```

**Expected behavior**

`Codec` value is codec

**Actual behavior**

`Codec` value is null

**Known workarounds**

Do not use `module rec`

**Related information**

* .NET 7
* Visual Studio 17.8 Preview 2.0

**Comment by vzarytovskii** [comment]
It was working in https://github.com/dotnet/fsharp/commit/30bbce6, however auto-bisecting it doesn't work for me for whatever reason, since it's always failing for me. 
**But**, if I checkout to https://github.com/dotnet/fsharp/commit/30bbce6 manually, it works.

**Comment by vzarytovskii** [comment]
Broken in https://github.com/dotnet/fsharp/commit/5483858

**Comment by vzarytovskii** [comment]
What we generate for binding (roughtly):
```csharp
.property class [Fleece]Fleece.Codec`4<class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class Domain.Assessee, class Domain.Assessee> Codec()
{
	.custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = (
		01 00 09 00 00 00 00 00
	)
	.get class [Fleece]Fleece.Codec`4<class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class Domain.Assessee, class Domain.Assessee> Domain.Events.AssessmentDataValidated/CodecsExtensions/AsseseeCodecs::get_Codec()
}

.method public specialname static 
	class [Fleece]Fleece.Codec`4<class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class Domain.Assessee, class Domain.Assessee> get_Codec () cil managed 
{
	// Method begins at RVA 0x2ce0
	// Header size: 1
	// Code size: 6 (0x6)
	.maxstack 8

	IL_0000: ldsfld class [Fleece]Fleece.Codec`4<class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class Domain.Assessee, class Domain.Assessee> '<StartupCode$ModuleRecIssueRepro>.$Domain.Events.AssessmentDataValidated'::'Codec@61-2'
	IL_0005: ret
} // end of method AsseseeCodecs::get_Codec
```

And a field:

```csharp
.field assembly static initonly class [Fleece]Fleece.Codec`4<class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class [Fleece.SystemTextJson]Fleece.SystemTextJson.Encoding, class Domain.Assessee, class Domain.Assessee> 'Codec@61-2'
.custom instance void [System.Runtime]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [System.Runtime]System.Diagnostics.DebuggerBrowsableState) = (
	01 00 00 00 00 00 00 00
)
```

I couldn't find the initialization code for the field.

**Comment by vzarytovskii** [comment]
Minimal repro:
[modulerecrepro.zip](https://github.com/dotnet/fsharp/files/12684554/modulerecrepro.zip)

**Comment by T-Gro** [comment]
Fixed by #16054
